pymobiledevice3/remote/remotexpc.py

```diff
@@ -0,0 +1,331 @@
+XpcUuid = Bytes(16)
+XpcFd = Int32ul
+XpcShmem = Struct('length' / Int32ul, Int32ul)
+XpcArray = Prefixed(Int32ul, LazyBound(lambda: XpcObject))
+
+XpcDictionaryEntry = Struct(
+    'key' / AlignedString,
+    'value' / LazyBound(lambda: XpcObject),
+)
+
+XpcDictionary = Prefixed(Int32ul, Struct(
+    'count' / Hex(Int32ul),
+    'entries' / If(this.count > 0, Array(this.count, XpcDictionaryEntry)),
+))
+
+XpcObject = Struct(
+    'type' / XpcMessageType,
+    'data' / Switch(this.type, {
+        XpcMessageType.DICTIONARY: XpcDictionary,
+        XpcMessageType.STRING: XpcString,
+        XpcMessageType.INT64: XpcInt64,
+        XpcMessageType.UINT64: XpcUInt64,
+        XpcMessageType.DOUBLE: XpcDouble,
+        XpcMessageType.BOOL: XpcBool,
+        XpcMessageType.NULL: XpcNull,
+        XpcMessageType.UUID: XpcUuid,
+        XpcMessageType.POINTER: XpcPointer,
+        XpcMessageType.DATE: XpcDate,
+        XpcMessageType.DATA: XpcData,
+        XpcMessageType.FD: XpcFd,
+        XpcMessageType.SHMEM: XpcShmem,
+        XpcMessageType.ARRAY: XpcArray,
+    }, default=Probe(lookahead=20)),
+)
+
+XpcPayload = Struct(
+    'magic' / Hex(Const(0x42133742, Int32ul)),
+    'protocol_version' / Hex(Const(0x00000005, Int32ul)),
+    'message' / XpcObject,
+)
+
+XpcWrapper = Struct(
+    'magic' / Hex(Const(0x29b00b92, Int32ul)),
+    'flags' / Default(XpcFlags, XpcFlags.ALWAYS_SET),
+    'size' / Hex(Int64ul),
+    'message_id' / Hex(Default(Int64ul, 0)),
+    'payload' / If(this.size > 0, FixedSized(this.size, XpcPayload)),
+)
+
+
+def _get_dict_from_xpc_object(xpc_object):
+    type_ = xpc_object.type
+
+    if type_ == XpcMessageType.DICTIONARY:
+        if xpc_object.data.count == 0:
+            return {}
+        result = {}
+        for entry in xpc_object.data.entries:
+            result[entry.key] = _get_dict_from_xpc_object(entry.value)
+        return result
+
+    elif type_ == XpcMessageType.ARRAY:
+        result = []
+        for entry in xpc_object.data.entries:
+            result.append(_get_dict_from_xpc_object(entry.value))
+        return result
+
+    elif type_ == XpcMessageType.BOOL:
+        return bool(xpc_object.data)
+
+    elif type_ == XpcMessageType.INT64:
+        return XpcInt64Type(xpc_object.data)
+
+    elif type_ == XpcMessageType.UINT64:
+        return XpcUInt64Type(xpc_object.data)
+
+    elif type_ == XpcMessageType.UUID:
+        return uuid.UUID(bytes=xpc_object.data)
+
+    elif type_ in (XpcMessageType.STRING, XpcMessageType.DATA):
+        return xpc_object.data
+
+    else:
+        raise Exception(f'deserialize error: {xpc_object}')
+
+
+class XpcInt64Type(int):
+    pass
+
+
+class XpcUInt64Type(int):
+    pass
+
+
+def get_object_from_xpc_wrapper(payload: bytes):
+    payload = XpcWrapper.parse(payload).payload
+    if payload is None:
+        return None
+    return _get_dict_from_xpc_object(payload.message)
+
+
+def build_xpc_object(payload) -> Mapping:
+    if isinstance(payload, list):
```

Comment:
Perhaps a dictionary will be more open closed than if-else?

Reply:
done